1.md

```diff
@@ -0,0 +1,200 @@
+  ability to instantiate new module graphs when dependencies change and the
+  ability to bequeath subgraphs to new graphs.
+  * Node.js [defers][node-hmr] to ECMAScript to provide a module loader
+    interface to aid HMR.
+* persistent testing apparatuses ([Jest][jest]), because a persistent service
+  reinstantiates whole module graphs to reconstruct tests and test subjects.
+  * Jest currently resorts to exploiting Node.js's [vm][vm-context] module to
+    instantiate separate realms and attempts ([and
+    fails][jest-ses-interaction]) to provide the illusion of a single realm by
+    patching client realms with some of the intrinsics of the host realm.
+
+## Examples
+
+### Analyzing a module graph
+
+The following code produces a module graph from modules plainly published on
+the web using URLs as import specifiers and memo keys.
+No modules are executed.
+
+```js
+const graph = new Map();
+
+const load = async url => {
+  if (graph.has(url)) {
+    return;
+  }
+  const response = await fetch(url);
+
+  // Account for redirects.
+  if (response.url !== url) {
+    graph.set(url, new Set([response.url]));
+    return load(response.url);
+  }
+
+  const edges = new Set();
+  graph.set(url, edges);
+
+  const text = await response.text();
+  const source = new ModuleSource(text);
+
+  const dependencies = [];
+  for (const binding of source.bindings) {
+    if ('from' in binding) {
+      const importUrl = new URL(binding.from, url).href;
+      edges.add(importUrl);
+      dependencies.push(load(importUrl));
+    }
+  }
+  await Promise.all(dependencies);
+};
+
+await load('https://example.com/example.js');
+```
+
+## Design Questions
+
+Do we also need to reflect `isAsync`?
+This appears to depend on whether implementations need to know
+whether execution will be asynchronous before actually beginning to execute.
+XS appears to have managed to implement virtual module sources without
+an explicit indicator.
+ECMA-262 currently has [[isAsync]] on ***Cyclic Module Record***,
+which would suggest that, if engines can be implemented without knowing
+a source will be asynchronous, the specification will need to be refactored to
+reflect that.
+
+## Design Rationales
+
+The property `needsImportMeta` allows virtual import hooks to omit properties
+from the `importMeta` of any `Module` instance derived from the source,
+having proof that the module will never access `import.meta`.
+Concretely, `import.meta.resolve` would be a closure over the module's referrer
+in hosts that provide it.
+In module graphs with thousands of module instances that largely do not use
+this property, avoiding the allocation of per-module closures can allow a
+significant reduction in memory pressure.
+
+A similar optimization might be possible for `import`.
+With the design as written, `needsImport` would only be `false` for modules
+that make no use of static `import` or `export` `from` clauses and also never
+use the syntactic form for dynamic `import`.
```

Comment:
I thought `needsImport` means `needs dynamic import`. Static imports and exports are resolved without any call to the `context.import()`.

Each module will need to have its own `context.import` function (in order to distinguish the import referrer) so having this property also allows to save more memory.

Reply:
With @caridy’s design, we use `[[Context]].[[Evaluators]].[[ImportHook]]` for both static and dynamic import, and the ImportHook receives the importMeta, so a single `importHook` can serve a large module graph.

The virtual module source chapter does end up reading `needsImport`, but that is indeed a case where the import function needs to be bound to the referrer and only necessary for dynamic import. All the static imports on the namespace objects before calling `execute` / `initialize`.